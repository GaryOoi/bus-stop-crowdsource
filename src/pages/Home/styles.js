import styled from "styled-components";

export const Section = styled.div`
  padding: 3em 0;
  position: relative;
  @media (max-width: 767.98px) {
    padding-top: 4em;
  }
`;

export const SectionHeader = styled.h2`
  font-size: 30px;
  font-weight: 800;
`;

export const SpinnerWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;
