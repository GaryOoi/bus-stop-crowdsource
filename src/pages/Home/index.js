import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Container, Spinner } from "react-bootstrap";

import {
  getBusstopsRequest,
  getBusstopsSuccess,
  getBusstopsFail,
} from "actions/busstopActions";
import {
  getBusstops,
  getBusstopsLoading,
  getBusstopsError,
} from "selectors/busstopSelectors";

import MessageModal from "components/MessageModal";
import { getData } from "utils/request";
import BusCard from "components/BusCard";
import { Section, SectionHeader, SpinnerWrapper } from "./styles";

function Home() {
  const dispatch = useDispatch();
  const busStops = useSelector(getBusstops);
  const busstopsLoading = useSelector(getBusstopsLoading);
  const busstopsError = useSelector(getBusstopsError);
  const [modelData, setModelData] = useState({
    visibility: false,
    message: "",
  });

  const handlePressModal = () => {
    setModelData({
      ...modelData,
      visibility: !modelData.visibility,
    });
  };

  useEffect(() => {
    const fetchAPI = async () => {
      try {
        dispatch(getBusstopsRequest());
        const result = await getData("busstops");
        dispatch(getBusstopsSuccess(result));
      } catch (err) {
        dispatch(getBusstopsFail(err));
        setModelData({
          visibility: true,
          message: err,
        });
      }
    };
    if (busStops.length < 1) {
      fetchAPI();
    }
  }, [busStops.length, dispatch]);

  return (
    <>
      <Section>
        {busstopsLoading || busstopsError !== "" ? (
          <SpinnerWrapper>
            <Spinner animation="border" role="status">
              <span className="sr-only">Loading...</span>
            </Spinner>
          </SpinnerWrapper>
        ) : (
          <Container>
            <SectionHeader className="mb-5">All Bus Stops</SectionHeader>
            {busStops.map((busstop) => {
              return <BusCard key={busstop.id} busstop={busstop} />;
            })}
          </Container>
        )}
      </Section>
      <MessageModal modelData={modelData} onClick={handlePressModal} />
    </>
  );
}

export default Home;
