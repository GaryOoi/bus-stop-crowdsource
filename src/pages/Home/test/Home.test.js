import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";

import Home from "../index";

test("Test home page", () => {
  const mockStore = configureStore([]);
  const store = mockStore({
    busStop: {
      isLoading: false,
      error: "",
      busStops: [
        {
          id: 1,
          name: "xyz",
          currentAmount: 6,
          totalAmount: 700,
        },
      ],
    },
  });

  const { container } = render(
    <Provider store={store}>
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    </Provider>
  );

  expect(container.firstChild).toMatchSnapshot();
});
