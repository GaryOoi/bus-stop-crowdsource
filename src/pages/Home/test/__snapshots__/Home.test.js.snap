// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div
  class="sc-AxhUy eoRDWV"
>
  <div
    class="container"
  >
    <h2
      class="sc-AxgMl chuQjx mb-5"
    >
      All Bus Stops
    </h2>
    <div
      class="mb-4 card"
    >
      <div
        class="sc-AxjAm DJJLR d-flex align-items-center card-body"
      >
        <div>
          <h5
            class="mb-3"
          >
            xyz
          </h5>
          <div
            class="sc-AxirZ lfktq"
          >
            Funding Target: $
            700
          </div>
          <div
            class="sc-AxiKw eixFbw"
          >
            0.86
            % funded
          </div>
        </div>
        <button
          class="sc-AxhCb igDhVv btn btn-primary"
          type="button"
        >
          Donate Now
        </button>
      </div>
    </div>
  </div>
</div>
`;
